### ISSUE TITLE
SEVERE Performance issue

### ISSUE THREAD

**Issue by GlassedSilver** [Tzahi12345/YoutubeDL-Material#264]
I am using Firefox on Windows 10 and when I load the homepage where YTDL-M will load ALL of my downloaded videos' thumbnails unpaginated (which I would always prefer over lazy-loading, no idea how pop-in got trendy when everything else these days moves towards the speedy experience... guess the unscrolled first impression is what counts more than anything else these days...).

This is obviously a problem when you've got a lot of downloads. :O

Now my YTDL-M instance purely runs on spinning rust, sure, but...
Where else do you put all of your many many dozens, maybe hundreds of GBs of video downloads, right?

So now I get the "A web page is slowing down your browser. What would you like to do?" message repeatedly even after clicking on "wait".

This will only become worse as time (and downloads) goes on.

I say we need to load paginated, as much as I like the ability to ctrl-F for everything, we do have the little search built-in, so as long as that will search through everything (even on other pages than the one currently loaded) then I'd rather have my fast loading home page back. Right now I'm just dreading to download beyond my automated subscriptions. Let alone browse for fun.

Another system to present the videos would come eventually anyhow I guess once we consider to implement smart suggestions/recommendations YT-style. (as in... sometimes you may be on a spree for a certain kind of videos and then you get recommended stuff based on tags, description keywords, channels, category, etc... I digress)

Any thoughts? I can't be the only power user experiencing this. :O

**Comment by GlassedSilver** [comment]
Severe as in: the page load needs literal minutes by now... No, I'm not kidding.

**Comment by Tzahi12345** [comment]
Can you try disabling `Include thumbnail` in the Downloader tab in settings? I've noticed that with lots of videos it takes forever to send over all thumbnails depending on the internet connection.

Related issue: #225

**Comment by GlassedSilver** [comment]
Internet connection? My server is running in my own LAN over 1Gbit ethernet.
Loading all the thumbnails itself is the problem.

And wait, are you suggesting not to download the thumbnails? I do want to preserve the thumbnails of videos locally though. :/

**Comment by Tzahi12345** [comment]
The way that the files (audio/video/subscriptions) work right now is that if `Include thumbnail` is enabled, thumbnails are sent alongside metadata. Otherwise, the external URL for the thumbnail is sent over.

I'm now realizing this may be stupid, because it takes the getAllFiles route *forever* to send that over. I should instead have a route that the frontend uses to get the thumbnail from the backend, and let the UI deal with the loading process (it'll be done lazily, which is ideal). While writing this comment I decided to go ahead and make those changes ([see commit here](https://github.com/Tzahi12345/YoutubeDL-Material/commit/2f541a49dfa961501e73aac027ba3e1cd588e7ab)), in about an hour the `nightly` should be updated and include this fix! Let me know if that works for you, if so problem solved :)

**Comment by GlassedSilver** [comment]
Could it be that the nightly docker build is still an older one?

**Comment by Tzahi12345** [comment]
It sounds unlikely. Perhaps you need to delete your local image of the `nightly`, by deleting your container (you might want to back up your volumes for this one) and then deleting the image. I'll include the instructions below if needed:

1. `docker ps -a` to find your container ID
2. `docker rm {container ID}` to delete the container
3. `docker images` to find the image ID for the nightly
4. `docker rmi {image ID}` to delete the image

Let me know if it still has issues. If so, if you don't mind sharing your instance, message me the details over reddit. If not, that's totally OK too, I understand!

**Comment by GlassedSilver** [comment]
Well if you look [here](https://hub.docker.com/r/tzahi12345/youtubedl-material/builds) it does seem like DH has last seen an update to nightly 10 days ago. hmm

**Comment by Tzahi12345** [comment]
Those are the auto builds I believe, they have been turned off. If you look [here](https://hub.docker.com/r/tzahi12345/youtubedl-material/tags) you can see the tag has been updated pretty recently.

But I do wonder if for some reason that's preventing you from updating. Can you check the creation date of your downloaded image? [See here for instructions.](https://stackoverflow.com/a/32705298/8088021)

**Comment by GlassedSilver** [comment]
> 
> 
> Those are the auto builds I believe, they have been turned off. If you look [here](https://hub.docker.com/r/tzahi12345/youtubedl-material/tags) you can see the tag has been updated pretty recently.
> 
> But I do wonder if for some reason that's preventing you from updating. Can you check the creation date of your downloaded image? [See here for instructions.](https://stackoverflow.com/a/32705298/8088021)

Seems pretty recent, but the webpage simply won't load.

Maybe this filling my log is related, but idk, seems like a ytdl, not a ytdl-m issue about trying and failing to download videos through my subscriptions?

`ERROR: Unable to download webpage: HTTP Error 429: Too Many Requests (caused by HTTPError()); please report this issue on https://yt-dl.org/bug . Make sure you are using the latest version; type youtube-dl -U to update. Be sure to call youtube-dl with the --verbose flag and include its complete output.`

Cheers

**Comment by Tzahi12345** [comment]
Ah the classic 429 error. I should make a page dedicated to this error and how to solve it. In another issue I [outlined my recommendations](https://github.com/Tzahi12345/YoutubeDL-Material/issues/166#issuecomment-659086304) -- if you can't use a proxy, you'll probably have to wait a few days until YouTube unblocks your IP (that's the fundamental issue here). If you can [use a proxy](https://github.com/Tzahi12345/YoutubeDL-Material/issues/177#issuecomment-722685712), you're in luck!

>Seems pretty recent, but the webpage simply won't load.

Are you saying the app won't load or downloaded videos fail? We may have another issue here (someone else had this problem)

**Comment by GlassedSilver** [comment]
I'm saying I can look at a dark blank page for minutes with the site title loading, but Firefox loads and loads and loads and loads... Basically same issue as before.

How would a proxy solve it btw? If I route that many requests at once through the proxy, won't the proxy URL get banned instead then?
And how can I still load the YT website on my browser using the same IP after all? I thought ytdl-m doesn't access through the API?

**Comment by GlassedSilver** [comment]
Ok apparently after a restart of the container and a bit of luck it does load the page now, but those loading animations are stuttery as heck.

Still think pagination and therefore a limitation of db calls would help here... ?

Not sure though. Can't make this instance WAN-facing btw, but I think it should be re-creatable.

**Comment by Tzahi12345** [comment]
>How would a proxy solve it btw? If I route that many requests at once through the proxy, won't the proxy URL get banned instead then?
>I'm saying I can look at a dark blank page for minutes with the site title loading, but Firefox loads and loads and loads and loads... Basically same issue as before.

Weird, maybe it gets stuck in some loop or intensive tasks. Hard to know why... If that's the case, it would make sense why restarting fixes it. If it happens again, try looking at the network tab in the browser dev tools to see which request is taking forever (or if it's all of them).

>And how can I still load the YT website on my browser using the same IP after all? I thought ytdl-m doesn't access through the API?

It won't block access to the YT website, I guess they got smart about just blocking downloads. It's why I suspect using cookies could resolve it but I'm not sure.

>Still think pagination and therefore a limitation of db calls would help here... ?

Yes -- ish. Remote pagination would be great, but the amount of data we're getting with the list of videos isn't huge (it is huge in the version you're using, but not currently). The most intensive part of the process is rendering the file cards, so local pagination would probably get us 90% of the way with reducing loading times/lag. So yes, reducing DB calls would help, but we can probably get away with avoiding listing *all* of the videos at once (which is relatively easy to implement).

>Not sure though. Can't make this instance WAN-facing btw, but I think it should be re-creatable.

I see, yeah all you need to do is copy the folders and you have a cloned instance. It would certainly help with debugging here. I assume deleting the container & images didn't help? I understand if you're scared to do that whole process, I'd be too lol

**Comment by GlassedSilver** [comment]
Not fixed, page itself is loaded, but no thumbnails get loaded and eventually the page just keeps staying in a frozen state, can't even scroll, etc...
It's hard-locked.

Probably the suggestion you have on your mind would help yes.

Yeah the thing is I won't open up anything on my LAN-only server to be served publicly.
Not even with VPN, because before I consider that I have to read up more on network security for sure.
Also I know that unRAID is made as an OS that assumes mostly anything to be trustable

**Comment by Tzahi12345** [comment]
I actually love the paranoia, it's by far the better extreme versus not caring at all. A VPN would be as secure as the private key, so not bad but if it gets out it's basically useless (not really a *private* network anymore). Every time I've tried to set one up it's been a pain but I'm not a sysadmin so that probably explains it.

>Not fixed, page itself is loaded, but no thumbnails get loaded and eventually the page just keeps staying in a frozen state, can't even scroll, etc...
>It's hard-locked.

Agh, I wish I could somehow see what's going on under the hood. Ultimately I think if you can get the nightly up to date I think it'll be fixed. Does RAM or CPU usage spike? Anything weird in the browser console or network tab? (right click -> inspect element -> console/network)

**Comment by GlassedSilver** [comment]
Could you provide a way to read out the version of the installed instance without having to fumble within the docker image?

Cause the creation date seems fine, but either I'm still on the old version or it's simply not fixed.

I can't even load the network tab, because my Firefox comes to a crawl so much.

**Comment by GlassedSilver** [comment]
So I started the network logger before loading it up and now the page doesn't load at the moment to the better than blank state, but have this anyhow?

![image](https://user-images.githubusercontent.com/1912133/100809434-63ab9680-3436-11eb-88be-f497a58871d1.png)

**Comment by Tzahi12345** [comment]
>Could you provide a way to read out the version of the installed instance without having to fumble within the docker image?

Hm, I can imagine this could be done in the build process. One way you could tell right now is by going into the container and opening the `app/package.json` and seeing if `axios` is in there somewhere. If so you're on the latest version.

>So I started the network logger before loading it up and now the page doesn't load at the moment to the better than blank state, but have this anyhow?

That is so weird. Can you try opening it up in an incognito tab? Make sure you use the base URL with nothing else added (`192.168.2.107:8999`) because perhaps when you try going to the home page there's some kind of redirect loop involved.

**Comment by GlassedSilver** [comment]
Yes, `axios` is in there.

I think at this point in time the issue isn't network traffic because I also don't see peaks in the server's CPU usage.
RAM is a-ok as well. In fact nothing is noticeable on the server end.

The CLIENT however... Yeah I think it's definitely a browser-faced issue. The browser is tasked to do too much at once.

I also tested this across multiple browsers.
Chrome for Windows
Firefox for Windows
Firefox for Android

The latest versions respectively and also a bit older ones, so it's definitely not the browser's fault.

Another thing to note:

Even if we wanted to keep loading all of these items at once, just placing the playlists at the bottom means you WILL have to scroll to the end to get to them.
This is obviously not gonna age well even with moderate usage. :o

**Comment by Tzahi12345** [comment]
I've gone ahead and added a paginator to the home page, let me know if this fixes the issue!

**Comment by GlassedSilver** [comment]
Sometimes I still get the behavior, but I think that's cached scripts maybe? Not sure.

Even in Privacy Mode I still get it, but it's considerably better.

Next time it happens I'll look into the network details tab again!

**Comment by GlassedSilver** [comment]
MAYBE this popping up in the console has something to do with it?

```
Script terminated by timeout at:
_balanceNamespaceList@http://192.168.2.107:8999/1-es2015.9f67c55e149357ac491a.js:1:82186
createNamespace@http://192.168.2.107:8999/1-es2015.9f67c55e149357ac491a.js:1:81978
register@http://192.168.2.107:8999/1-es2015.9f67c55e149357ac491a.js:1:82492
register@http://192.168.2.107:8999/1-es2015.9f67c55e149357ac491a.js:1:96083
createRenderer@http://192.168.2.107:8999/1-es2015.9f67c55e149357ac491a.js:1:109250
Io@http://192.168.2.107:8999/main-es2015.b33803ca3b13aa9213b7.js:1:60128
fo/<@http://192.168.2.107:8999/main-es2015.b33803ca3b13aa9213b7.js:1:55254
fo@http://192.168.2.107:8999/main-es2015.b33803ca3b13aa9213b7.js:1:55345
Si@http://192.168.2.107:8999/main-es2015.b33803ca3b13aa9213b7.js:1:84598
template@http://192.168.2.107:8999/1-es2015.9f67c55e149357ac491a.js:1:1086066
ao@http://192.168.2.107:8999/main-es2015.b33803ca3b13aa9213b7.js:1:54884
io@http://192.168.2.107:8999/main-es2015.b33803ca3b13aa9213b7.js:1:53221
Vo@http://192.168.2.107:8999/main-es2015.b33803ca3b13aa9213b7.js:1:61206
io/<@http://192.168.2.107:8999/main-es2015.b33803ca3b13aa9213b7.js:1:53424
io@http://192.168.2.107:8999/main-es2015.b33803ca3b13aa9213b7.js:1:53433
createEmbeddedView@http://192.168.2.107:8999/main-es2015.b33803ca3b13aa9213b7.js:1:68677
createEmbeddedView@http://192.168.2.107:8999/main-es2015.b33803ca3b13aa9213b7.js:1:69651
_applyChanges/<@http://192.168.2.107:8999/main-es2015.b33803ca3b13aa9213b7.js:1:253557
forEachOperation@http://192.168.2.107:8999/main-es2015.b33803ca3b13aa9213b7.js:1:102080
_applyChanges@http://192.168.2.107:8999/main-es2015.b33803ca3b13aa9213b7.js:1:253476
ngDoCheck@http://192.168.2.107:8999/main-es2015.b33803ca3b13aa9213b7.js:1:253428
zt@http://192.168.2.107:8999/main-es2015.b33803ca3b13aa9213b7.js:1:33727
Ut@http://192.168.2.107:8999/main-es2015.b33803ca3b13aa9213b7.js:1:33557
Ht@http://192.168.2.107:8999/main-es2015.b33803ca3b13aa9213b7.js:1:33283
uo@http://192.168.2.107:8999/main-es2015.b33803ca3b13aa9213b7.js:1:53740
uo/<@http://192.168.2.107:8999/main-es2015.b33803ca3b13aa9213b7.js:1:53894
uo@http://192.168.2.107:8999/main-es2015.b33803ca3b13aa9213b7.js:1:53948
Ro@http://192.168.2.107:8999/main-es2015.b33803ca3b13aa9213b7.js:1:61055
uo/<@http://192.168.2.107:8999/main-es2015.b33803ca3b13aa9213b7.js:1:54428
uo@http://192.168.2.107:8999/main-es2015.b33803ca3b13aa9213b7.js:1:54437
uo/<@http://192.168.2.107:8999/main-es2015.b33803ca3b13aa9213b7.js:1:53894
uo@http://192.168.2.107:8999/main-es2015.b33803ca3b13aa9213b7.js:1:53948
Ro@http://192.168.2.107:8999/main-es2015.b33803ca3b13aa9213b7.js:1:61055
uo/<@http://192.168.2.107:8999/main-es2015.b33803ca3b13aa9213b7.js:1:54428
uo@http://192.168.2.107:8999/main-es2015.b33803ca3b13aa9213b7.js:1:54437
uo/<@http://192.168.2.107:8999/main-es2015.b33803ca3b13aa9213b7.js:1:53894
uo@http://192.168.2.107:8999/main-es2015.b33803ca3b13aa9213b7.js:1:53948
Ro@http://192.168.2.107:8999/main-es2015.b33803ca3b13aa9213b7.js:1:61055
uo/<@http://192.168.2.107:8999/main-es2015.b33803ca3b13aa9213b7.js:1:54428
uo@http://192.168.2.107:8999/main-es2015.b33803ca3b13aa9213b7.js:1:54437
co@http://192.168.2.107:8999/main-es2015.b33803ca3b13aa9213b7.js:1:54761
Bo/<@http://192.168.2.107:8999/main-es2015.b33803ca3b13aa9213b7.js:1:61597
Bo@http://192.168.2.107:8999/main-es2015.b33803ca3b13aa9213b7.js:1:61617
detectChanges@http://192.168.2.107:8999/main-es2015.b33803ca3b13aa9213b7.js:1:68079
tick@http://192.168.2.107:8999/main-es2015.b33803ca3b13aa9213b7.js:1:172271
next/<@http://192.168.2.107:8999/main-es2015.b33803ca3b13aa9213b7.js:1:170710
invoke@http://192.168.2.107:8999/polyfills-es2015.5b408f108bcea938a7e2.js:1:54174
onInvoke@http://192.168.2.107:8999/main-es2015.b33803ca3b13aa9213b7.js:1:161453
invoke@http://192.168.2.107:8999/polyfills-es2015.5b408f108bcea938a7e2.js:1:54114
run@http://192.168.2.107:8999/polyfills-es2015.5b408f108bcea938a7e2.js:1:49586
run@http://192.168.2.107:8999/main-es2015.b33803ca3b13aa9213b7.js:1:162109
next@http://192.168.2.107:8999/main-es2015.b33803ca3b13aa9213b7.js:1:170696
subscribe/r<@http://192.168.2.107:8999/main-es2015.b33803ca3b13aa9213b7.js:1:151173
__tryOrUnsub@http://192.168.2.107:8999/main-es2015.b33803ca3b13aa9213b7.js:1:5040
next@http://192.168.2.107:8999/main-es2015.b33803ca3b13aa9213b7.js:1:4261
_next@http://192.168.2.107:8999/main-es2015.b33803ca3b13aa9213b7.js:1:3432
next@http://192.168.2.107:8999/main-es2015.b33803ca3b13aa9213b7.js:1:3206
next@http://192.168.2.107:8999/main-es2015.b33803ca3b13aa9213b7.js:1:12041
emit@http://192.168.2.107:8999/main-es2015.b33803ca3b13aa9213b7.js:1:151044
Lf@http://192.168.2.107:8999/main-es2015.b33803ca3b13aa9213b7.js:1:162498
onHasTask@http://192.168.2.107:8999/main-es2015.b33803ca3b13aa9213b7.js:1:161597
hasTask@http://192.168.2.107:8999/polyfills-es2015.5b408f108bcea938a7e2.js:1:55065
_updateTaskCount@http://192.168.2.107:8999/polyfills-es2015.5b408f108bcea938a7e2.js:1:55301
_updateTaskCount@http://192.168.2.107:8999/polyfills-es2015.5b408f108bcea938a7e2.js:1:51656
runTask@http://192.168.2.107:8999/polyfills-es2015.5b408f108bcea938a7e2.js:1:50399
v@http://192.168.2.107:8999/polyfills-es2015.5b408f108bcea938a7e2.js:1:56786
invokeTask@http://192.168.2.107:8999/polyfills-es2015.5b408f108bcea938a7e2.js:1:55871
h@http://192.168.2.107:8999/polyfills-es2015.5b408f108bcea938a7e2.js:1:68125
p@http://192.168.2.107:8999/polyfills-es2015.5b408f108bcea938a7e2.js:1:68371
```

idk, timeout just seems very interesting.

**Comment by Tzahi12345** [comment]
I've done two things which should improve performance:

1. Update to Angular 11 ([commit](https://github.com/Tzahi12345/YoutubeDL-Material/commit/3f10986cdfca31af0dd7c7bf8c80bd1ef656d719))
2. Performance optimization during loading which maxes out the number of ghost cards to 10 ([commit](https://github.com/Tzahi12345/YoutubeDL-Material/commit/ff1bb8dee192e833775a0b80abf62bb3ff792d52))

Let me know if this fixes it for you!

**Comment by GlassedSilver** [comment]
Seems fixed!
Thank you so much for this! YTDL-M is way snappier now, also did go ahead and put the container on an SSD now.
Adding SSDs for the applications on my server was one of the best tech decisions I've made so far this year! :) (after building a new PC that is and NOT waiting for the ever not available RTX30 series cards. Guess I'll just go straight to its successor once my 2070S shows its age. ;) )

**Comment by Tzahi12345** [comment]
I'm currently trying to pick up a 3070 (upgrading from a 1060). I doubt I'll get my hands on that for a whiiile lmao

Glad it's fixed, I'll close this issue :)
